### PR TITLE
Generate for klu, amd, colamd, ccolamd

### PR DESCRIPTION
**Pull request by rayegun**
This adds generation for some of the missing libraries from SuiteSparse, including KLU. 

Version numbers are incorrect because the most recent SuiteSparse_jll are restricted to 1.7, and I cannot get Clang.jl working on 1.7 yet. Once that happens this should be mostly ready and then I can work on Julian wrappers. 

It looks like @Gnimuc did the first generation. Are you able to run this generator with Clang on 1.7? Relatedly, can we do the Clang generation in CI?

The most important question, however, is whether this should be done in a new package (in JuliaSparse for instance) rather than here.

https://github.com/JuliaSmoothOptimizers/AMD.jl already wraps COLAMD and AMD, although they do not yet use the jll. 

I see a couple options:

1. Merge into this repo, expose the new functionality however necessary (as part of the stdlib or this package if it becomes one).
2. Fork and create JuliaSparse/SuiteSparse.jl, release all functionality there.
3. Keep functionality separate. What's already in SuiteSparse.jl stays there, perhaps using the Clang generated code; COLAMD and AMD generated code goes to JuliaSmoothOptimizers; KLU code goes to JuliaSparse or SciML or somewhere else.

KLU was requested by Chris Rackauckas in https://github.com/SciML/LinearSolve.jl/issues/9. Also note, I would like to integrate these with SuiteSparseGraphBLAS.jl.

cc @ViralBShah

### CODE REVIEW

**Comment by rayegun** [comment]
Can someone approve workflow to see if this works? Also what are the thoughts on those 3 options above?

**Comment by ViralBShah** [comment]
I would personally be ok with providing all of these bindings in SuiteSparse.jl - it is very little new code and makes it easy for other packages to use SuiteSparse. @KristofferC @staticfloat may have some thoughts, in case this doesn't play well with the julia build.

**Comment by rayegun** [comment]
Do you want the higher level Julia wrapper (above the Clang generated one) here also?

**Comment by ViralBShah** [comment]
For higher level user facing APIs, they have to be carefully designed and reviewed, but yes, they can go in here as well.

**Comment by rayegun** [comment]
I'll get started on the higher level wrapper then in a different PR, done probably in a week or two. What is the 1.8 merge deadline?

**Comment by KristofferC** [comment]
> What is the 1.8 merge deadline?

Feature freeze date hasn't been announced yet.

**Comment by StefanKarpinski** [comment]
But typically 4 months after the 1.7 feature freeze happened, whenever that was.

**Comment by rayegun** [comment]
June 1? So I've got two months to get this out. Much like 3.5 roentgen, not great, not terrible.

**Comment by StefanKarpinski** [comment]
You don't really have to think too much about feature freezes. The main branch is always open, just merge it when it's done and it will be in the following release.

**Comment by ViralBShah** [comment]
Question about clang generation in CI may be for @DilumAluthge. For now, we had decided to generate externally and check in (so that we didn't have to wait for making changes to our CI infrastructure). But it would be desirable to do it in CI. 

Perhaps @Wimmerer can also familiarize with the CI with input from @DilumAluthge and @staticfloat.

**Comment by ViralBShah** [comment]
@Wimmerer Is this otherwise ready for merge? It seems to add more ccalls, but shouldn't break anything existing. Is that right? 

Given the dependency on Julia 1.7, perhaps the ideal thing to do is to merge after 1.7 is released, just in case we need to bump SuiteSparse for the Julia 1.7 release.

**Comment by rayegun** [comment]
Let me look it over before merge but yes it should be ready, I'll take a look tomorrow. On a road trip at the moment. I'll touch base with them tomorrow.

File: gen/Manifest.toml
Comment (Gnimuc) [inline gen/Manifest.toml]: Looks like Julia-1.7-beta4 will ship LLVM-12.0.1, so we may need to update this.

File: gen/generator.jl
Comment (Gnimuc) [inline gen/generator.jl]: Is it intended to not include `amd_h`, `colamd_h` and `ccolamd_h`? Perhaps, `klu_h` includes these headers indirectly?

**Comment by ViralBShah** [comment]
@Wimmerer Can you review the comments above, and if you have the time, let's get this merged. I'd like to see if this can make it into 1.7. Not a big deal if it doesn't, since there'll always be a new release around the corner.

File: gen/Manifest.toml
Comment (rayegun) [reply gen/Manifest.toml]: This is just for generators right? Do those have to match any particular version?

File: gen/generator.jl
Comment (rayegun) [reply gen/generator.jl]: Unintentional, let's see what happens

**Comment by rayegun** [comment]
@Gnimuc Are you able to run the generator now? I get

```
julia> include("generator.jl")
[ Info: processing aarch64-apple-darwin20
┌ Error: Tree Hash Mismatch!
│   Expected git-tree-sha1:   fbb170fa64a06b66dd9a16aaa21bf12848042d9a
│   Calculated git-tree-sha1: 01305ea531fe03e26c82cdc27a625e2a45097caf
└ @ Pkg.Artifacts /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Artifacts.jl:364
ERROR: LoadError: AssertionError: failed to setup environment due to missing dirs, please file an issue at https://github.com/JuliaInterop/Clang.jl/issues.
Stacktrace:
 [1] get_system_dirs(triple::String, version::VersionNumber)
   @ Clang.JLLEnvs ~/.julia/packages/Clang/Ark1I/src/shards/JLLEnvs.jl:170
 [2] get_system_dirs
   @ ~/.julia/packages/Clang/Ark1I/src/shards/JLLEnvs.jl:101 [inlined]
 [3] get_default_args(triple::String)
   @ Clang.Generators ~/.julia/packages/Clang/Ark1I/src/generator/context.jl:230
 [4] top-level scope
   @ ~/SuiteSparse.jl/gen/generator.jl:48
 [5] include(fname::String)
   @ Base.MainInclude ./client.jl:451
 [6] top-level scope
   @ REPL[5]:1
```

even after fiddling with the environment. On Julia-1.7 beta 3

Also, what does https://github.com/JuliaLang/SuiteSparse.jl/blob/0f138c0df9fa4682f6bd73742fc1c83b4d34eacd/gen/generator.toml#L3 do? I haven't changed that.

File: gen/Manifest.toml
Comment (rayegun) [reply gen/Manifest.toml]: Actually it got removed by something? Not sure what happened

**Comment by Gnimuc** [comment]
> @Gnimuc Are you able to run the generator now? I get

Looks like the hash in the upstream has been changed. This can be fixed by manually removing the compile cache folder `.julia/compiled/Clang` and `using Clang` again.

> Also, what does
> 
> https://github.com/JuliaLang/SuiteSparse.jl/blob/0f138c0df9fa4682f6bd73742fc1c83b4d34eacd/gen/generator.toml#L3
> 
> do? I haven't changed that.

That is for configuring library names per header(controlled by regex). SuiteSparse has many library products, so this is needed.

**Comment by rayegun** [comment]
@Gnimuc I cannot get past the errors we discussed on several different devices. Are you able to just rerun the generator?

**Comment by Gnimuc** [comment]
Done!

**Comment by rayegun** [comment]
Ok this should be good to go. If we have to change something when higher-level documented interfaces are added we can do that, nothing depends on these yet, but I think I'd only add things like SLIP_LU at most.

**Review by Gnimuc — APPROVED** [review]

**Comment by ViralBShah** [comment]
So if we have done a good job with auto-generating all the right bindings, I suppose that `SuiteSparse.jl/src` should end up not having any `ccall`s, right? I notice one in cholmod and several in umfpack, which should hopefully be easy to replace with the wrapped ones.

**Comment by rayegun** [comment]
Yes, I can check on those and try to get them fixed ASAP, should be simple.

**Pull request merged**
